src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpHashJoin.java

```diff
@@ -0,0 +1,83 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import org.apache.jena.sparql.core.Var;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+import se.liu.ida.hefquin.engine.datastructures.SolutionMappingsIndex;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTable;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTableBasedOnOneVar;
+import se.liu.ida.hefquin.engine.datastructures.impl.SolutionMappingsHashTableBasedOnTwoVars;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultBlock;
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultElementSink;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+public class ExecOpHashJoin implements BinaryExecutableOp {
+    protected final SolutionMappingsIndex solMHashTableL;
+    protected Boolean child1InputComplete = false;
+
+    public ExecOpHashJoin(final ExpectedVariables inputVars1, final ExpectedVariables inputVars2) {
+        final Set<Var> joinVars = new HashSet<>( inputVars1.getCertainVariables());
+        joinVars.retainAll( inputVars2.getCertainVariables() );
+
+        if (joinVars.size() == 1 ){
+            final Var joinVar = joinVars.iterator().next();
+            this.solMHashTableL = new SolutionMappingsHashTableBasedOnOneVar(joinVar);
+        }
+        else if (joinVars.size() == 2){
+            final Iterator<Var> liVar = joinVars.iterator();
+            final Var joinVar1 = liVar.next();
+            final Var joinVar2 = liVar.next();
+
+            this.solMHashTableL = new SolutionMappingsHashTableBasedOnTwoVars(joinVar1, joinVar2);
+        }
+        else{
+            this.solMHashTableL = new SolutionMappingsHashTable(joinVars);
+        }
+    }
+
+    @Override
+    public int preferredInputBlockSize() {
+        return 1;
+    }
+
+    @Override
+    public boolean requiresCompleteChild1InputFirst() {
+        return true;
+    }
+
+    @Override
+    public void processBlockFromChild1( final IntermediateResultBlock input, final IntermediateResultElementSink sink, final ExecutionContext execCxt ) {
+        for ( final SolutionMapping smL : input.getSolutionMappings() ) {
+            solMHashTableL.add(smL);
+        }
+        wrapUpForChild1(sink, execCxt);
```

Comment:
No! You don't know whether this was the last call of 'processBlockFromChild1' or not. Only the code that uses this operator here knows, and that code takes care of calling 'wrapUpForChild1' in the end.